File: app/src/main/java/com/hover/stax/channels/UpdateChannelsWorker.java
package com.hover.stax.channels;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.work.Constraints;
import androidx.work.NetworkType;
import androidx.work.OneTimeWorkRequest;
import androidx.work.PeriodicWorkRequest;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import com.hover.stax.R;
import com.hover.stax.database.AppDatabase;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class UpdateChannelsWorker extends Worker {
	public final static String TAG = "UpdateChannelsWorker";

	public static String CHANNELS_WORK_ID = "CHANNELS";
	private final OkHttpClient client = new OkHttpClient();
	private final ChannelDao channelDao;
	private String errorMsg = null;

	public UpdateChannelsWorker(@NonNull Context context, @NonNull WorkerParameters params) {
		super(context, params);
		channelDao = AppDatabase.getInstance(context).channelDao();
	}

	public static PeriodicWorkRequest makeToil() {
		return new PeriodicWorkRequest.Builder(UpdateChannelsWorker.class, 24, TimeUnit.HOURS)
			       .setConstraints(netConstraint())
			       .build();
	}

	public static OneTimeWorkRequest makeWork() {
		return new OneTimeWorkRequest.Builder(UpdateChannelsWorker.class)
			       .setConstraints(netConstraint())
			       .build();
	}

	public static Constraints netConstraint() {
		return new Constraints.Builder().setRequiredNetworkType(NetworkType.CONNECTED).build();
	}

	@Override
	public Worker.Result doWork() {
		try {
			Log.v(TAG, "Downloading channels...");
			JSONObject channelsJson = downloadChannels(getUrl());
			JSONArray data = channelsJson.getJSONArray("data");
			for (int j = 0; j < data.length(); j++) {
				channelDao.insert(new Channel(data.getJSONObject(j).getJSONObject("attributes")));
			}
			Log.i(TAG, "Successfully downloaded and saved channels.");
			return Result.success();
		} catch (JSONException | NullPointerException e) {
			Log.e(TAG, "Error parsing channel data.", e);
			return Result.failure();
		} catch (IOException e) {
			Log.e(TAG, "Timeout downloading channel data, will try again.", e);
			return Result.retry();
		}
	}

	private String getUrl() {
		return getApplicationContext().getString(R.string.root_url) + getApplicationContext().getString(R.string.channels_endpoint);
	}

	private JSONObject downloadChannels(String url) throws IOException, JSONException {
		Request request = new Request.Builder().url(url).build();
		Response response = client.newCall(request).execute();
		JSONObject data = new JSONObject(response.body().string());
		return data;
//		return data.getJSONArray("attributes");
	}
}

File: app/src/main/java/com/hover/stax/channels/UpdateChannelsWorker.kt
/*
 * Copyright 2022 Stax
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.hover.stax.channels

import android.content.Context
import androidx.work.Constraints
import androidx.work.CoroutineWorker
import androidx.work.NetworkType
import androidx.work.OneTimeWorkRequest
import androidx.work.PeriodicWorkRequest
import androidx.work.WorkerParameters
import com.hover.stax.R
import com.hover.stax.database.channel.repository.ChannelRepository
import com.hover.stax.data.channel.ChannelRepositoryImpl
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject
import timber.log.Timber
import java.io.IOException
import java.util.concurrent.TimeUnit

class UpdateChannelsWorker(
    context: Context,
    params: WorkerParameters
) : CoroutineWorker(context, params), KoinComponent {

    private val client = OkHttpClient()
    private val channelRepository: ChannelRepository by inject()

    override suspend fun doWork(): Result {
        return try {
            val channelsJson = downloadChannels(url)
            if (channelsJson != null) {
                val data: JSONArray = channelsJson.getJSONArray("data")
                com.hover.stax.data.channel.ChannelRepositoryImpl.ChannelUtil.load(
                    data,
                    channelRepository,
                    applicationContext
                )
                Timber.v("Successfully Updated channels")
                Result.success()
            } else {
                Timber.d("Failed to update channels")
                Result.failure()
            }
        } catch (e: JSONException) {
            Result.failure()
        } catch (e: NullPointerException) {
            Result.failure()
        } catch (e: IOException) {
            Result.retry()
        }
    }

    private val url
        get() = applicationContext.getString(R.string.maathai_api_url)
            .plus(applicationContext.getString(R.string.channels_endpoint))

    private fun downloadChannels(url: String): JSONObject? {
        val request: Request = Request.Builder().url(url).build()
        val response: Response = client.newCall(request).execute()
        return response.body?.let { JSONObject(it.string()) }
    }

    companion object {
        const val CHANNELS_WORK_ID = "CHANNELS"
        const val TAG = "UpdateChannelsWorker"

        private val netConstraint =
            Constraints.Builder().setRequiredNetworkType(NetworkType.CONNECTED).build()

        fun makeToil(): PeriodicWorkRequest {
            return PeriodicWorkRequest.Builder(UpdateChannelsWorker::class.java, 7, TimeUnit.DAYS)
                .setConstraints(netConstraint)
                .build()
        }

        fun makeWork(): OneTimeWorkRequest {
            return OneTimeWorkRequest.Builder(UpdateChannelsWorker::class.java)
                .setConstraints(netConstraint)
                .build()
        }
    }
}